### ISSUE TITLE
Add `data source` support

### ISSUE THREAD

**Issue by vbauzysvmware** [terraform-providers/terraform-provider-vcd#218]
REF: https://github.com/terraform-providers/terraform-provider-vcd/pull/149
According to megalord:

> That's exactly what data sources are for. I'd encourage y'all to look into how other providers have used them (I can suggest AWS and DO from personal experience). Most resources that rely on other resources require the unfriendly id (specifically an arn in the AWS case) to be passed around.

> I'll go ahead and switch to the names, but you may keep this in mind. The more resources the provider defines, the worse this problem will get.

Links to documentation:
https://www.terraform.io/docs/configuration/data-sources.html
https://www.terraform.io/docs/configuration-0-11/data-sources.html

Such behaviour would add nice capabilities to provider to query resources which are not defined in terraform and link them together using IDs and not names. Names some times gives mentioned headaches and some resources like Independent disks allows disks with same name. This means some resources names are not unique.

**Comment by lvirbalas** [comment]
Would it be practical to create separate issues for different resources?
- Add Datasource for vApps
- Add Datasource for Catalogs and Catalog Items
- Add Datasource for Org VDC Networks

**Comment by vbauzysvmware** [comment]
What would be visible after POC for one of resources.

**Comment by dataclouder** [comment]
Here's an example of a data source implementation for `catalog_item`

```go
package vcd

import (
	"fmt"

	"github.com/hashicorp/terraform/helper/schema"
	"github.com/vmware/go-vcloud-director/v2/govcd"
)

func dataSourceVcdCatalogItem() *schema.Resource {
	return &schema.Resource{
		Read: dataSourceVcdCatalogItemRead,

		Schema: map[string]*schema.Schema{
			"name": &schema.Schema{
				Type:     schema.TypeString,
				Required: true,
			},
			"org": {
				Type:     schema.TypeString,
				Required: true,
			},
			"catalog": &schema.Schema{
				Type:     schema.TypeString,
				Required: true,
			},
			"description": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
			},
		},
	}
}

func dataSourceVcdCatalogItemRead(d *schema.ResourceData, meta interface{}) error {
	vcdClient := meta.(*VCDClient)

	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
	if err != nil {
		return fmt.Errorf(errorRetrievingOrg, err)
	}

	catalogName := d.Get("catalog").(string)
	catalogItemName := d.Get("name").(string)
	catalog, err := adminOrg.FindCatalog(catalogName)
	if err != nil {
		d.SetId("")
		return err
	}

	if catalog == (govcd.Catalog{}) {
		d.SetId("")
		return fmt.Errorf("unable to find catalog %s - empty catalog returned", catalogName)
	}
	catalogItem, err := catalog.FindCatalogItem(catalogItemName)
	if err != nil {
		d.SetId("")
		return err
	}
	if catalogItem == (govcd.CatalogItem{}) {
		d.SetId("")
		return fmt.Errorf("unable to find catalog item %s - empty catalog item returned", catalogItemName)
	}

	err = d.Set("description", catalogItem.CatalogItem.Description)
	if err != nil {
		d.SetId("")
		return err
	}
	d.SetId(catalogItem.CatalogItem.ID)
	// OR 
	// d.SetId(catalogItem.CatalogItem.HREF)
	return nil
}
```

The definition requires `org` and `catalog` and then the `Read` function sets the data source ID.

## How to use it?

The data source is used by getting the receiving resource ready to refer to it by its ID.

```hcl
data vcd_vapp_template "test" {
        name = "some_vm"
        catalog= "my_catalog"
        org = "my_org"
}

resource vcd_vm "test" {
    name = "test1"
    description = "terraform test VM basic config"
    template_id = "${data.vcd_vapp_template.test.id}"  # <<<<<
    power_on = true
    cpu_cores = 1
    memory = 1024
[...]
}
```
There are three important elements here:

1. The vm has a property `template_id` that refers to the catalog item (or vApp template item, which could be implemented separately, but this is a different story).
2. The `id` property of the data source is not listed its definition, but it exists by default. It's up to the data source implementation to fill it with the needed information. Depending on what API method is available to quickly retrieve the data source entity, the `id` could be filled with either the entity ID or the HREF.
3. There must be a function (preferably in `go-vcd-director` library) where we can get the data source entity (here: the catalog item) with a single query, using the information stored in the data source `id` field. This function will be used in the vm resource constructor to retrieve the data source entity and use it. In some cases, we don't even need to retrieve the entity. Case in point: to create a VM, we don't need to retrieve the data source itself, but only use its HREF; thus, if we have a data source where the `id` was filled with the HREF, we can just use it without further processing.

**Comment by vbauzysvmware** [comment]
Thank you Giuseppe, very good POC!

**Comment by buec011** [comment]
We would really like to see this. In many cases the data source has the same or similar name as the "create" resource, e.g. vcd_catalog_item for catalog_item datasource and resource.

https://www.terraform.io/docs/providers/aws/d/instance.html
https://www.terraform.io/docs/providers/aws/r/instance.html

I believe because you can reference similar output values, e.g. vcd_catalog_item.test.catalog vs data.vcd_catalog_item.test.catalog
(by the way, I'm missing Attributes Reference for resource vcd_catalog_item, but this is a new documentation bug...)

**Comment by buec011** [comment]
Maybe it will be helpful if I mention our demanded datasources here:

One is the example above, a data source for `vcd_catalog_item`. 
e.g. Input could be a Catalog Item Name (If the API has wildcard support, maybe with wildcard, abort with multiple matches, similar to https://www.terraform.io/docs/providers/aws/d/ami.html or vcd_catalog_items and save result in array, similar to https://www.terraform.io/docs/providers/aws/d/ami_ids.html )
```
data vcd_catalog_items myimage {
        name = "some_vm*"
}
```
and output/attributes catalog_name, template_name
```
resource "vcd_vapp_vm" "myinstance" {
  catalog_name  = data.vcd_catalog_items.myimage[0].catalog_name
  template_name = data.vcd_catalog_items.myimage[0].template_name
[...]
}
```

The second data source is `vcd_network_routed` or short `vcd_network`. Input Network name or cidr_block, e.g:
```
data vcd_network subnet {
   name = "mysubnet"
}
```
Output/attributes should be name, cidr_block similar to https://www.terraform.io/docs/providers/aws/d/subnet.html

`cidr_block = Networkadress/cidr` , e.g. `cidr_block=192.168.0.0/24`

```
output "subnet_cidr" {
  value = data.aws_subnet.subnet.cidr_block
}
```

**Comment by dataclouder** [comment]
@buec011 Thanks for your info, although in this case it comes a bit late, as the data source for catalog item has already been created (See [the docs](https://github.com/terraform-providers/terraform-provider-vcd/blob/master/website/docs/d/catalog_item.html.markdown)) but without any reference to the template name.
What we are doing with data sources is making the resource data available as DS. Thus, if the field was not in the resource, it would not end up in the data source.

What we need is an issue for each resource that is being converted, listing the properties that should be available as computed data.

**Comment by buec011** [comment]
> What we need is an issue for each resource that is being converted, listing the properties that should be available as computed data.

Ok, will open independent issues for my requirements

**Comment by dataclouder** [comment]
Implemented using the following PRs:
PR #333 (Networks and Edge Gateway)
PR #324 (VDC)
PR #322 (External Network)
PR #309 (catalog and catalog item)
PR #283 (Org)

**Comment by vbauzysvmware** [comment]
All new data source and imports are released and available in v2.5.0
